cis194/week1/daniel-deng/src/CIS194/Homework02/Exercise05.hs

```diff
@@ -0,0 +1,20 @@
+module CIS194.Homework02.Exercise05 (whatWentWrong) where
+
+import CIS194.Homework02.Log
+
+import Data.List
+
+isHighSeverityErr :: LogMessage -> Bool
+isHighSeverityErr (LogMessage (Error severity) _ _) = 50 <= severity
+isHighSeverityErr _                                 = False
+
+-- partial function, only used by `whatWentWrong`
+getErrTs :: LogMessage -> Int
+getErrTs (LogMessage (Error _) ts _) = ts
+
+-- partial function, only used by `whatWentWrong`
+getErrMsg :: LogMessage -> String
+getErrMsg (LogMessage (Error _) _ msg) = msg
+
+whatWentWrong :: [LogMessage] -> [String]
+whatWentWrong = map getErrMsg . sortOn getErrTs . filter isHighSeverityErr
```

Comment:
My question is - in this particular case, are partial functions really that eval? They are not exposed outside the module and we have taken care of the other cases with the filter at the first place. 

Isn't it true that we generally prefer `Prelude` functions over pattern matching and guard for list operations? Keen to hear your thoughts @shineli1984

Reply:
For this exercise, it's fine. However, in practice, we are at all costs to avoid partial functions. Even in the cases where a function is not exposed, others still need to come here to maintain this code base. They will not expect to see/consider non-exhaustive patterns. And normally we'll turn on non-exhaustive pattern matching warnings and turn them into errors to prevent compilation.
Also, many people turn on NoImplicitPrelude extension and use custom prelude like this one: http://hackage.haskell.org/package/protolude

Reply:
@shineli1984 ok that makes sense. Thanks for the in-depth explanation. 
Is something like this regarded as partial function? (Ignore the formatting for now)
```haskell
whatWentWrong :: [LogMessage] -> [String]
whatWentWrong logLine = map getMsg . sortOn getTs . filter isHighSeverityErr
                                              where isHighSeverityErr (LogMessage (Error sev) _ _) = 50 <= sev
                                                         isHighSeverityErr _ = False
                                                         getTs (LogMessage _ ts _) = ts
                                                         getMsg (LogMessage _ _ msg) = msg

```

Reply:
@shineli1984 thanks for the in-depth explanation. Here's my revision: https://github.com/fp-works/2019-winter-Haskell-school/pull/18/commits/ac2c25969e0957fc4c6aea13235761954f85de3b